### DISCUSSION TITLE
Why agent won't modify their own grpc class

### DISCUSSION THREAD

**Discussion by huange7** [open-telemetry/opentelemetry-java-instrumentation#10245] [Q&A] [answered]
There are instrumentation about grpc in java agent, and grpc is used in the agent code, will this grpc code be modified?
If not, by what means is it prevented?

**Comment by laurit** [comment, marked as answer]
Most of the agent code is loaded into separate class loader that is excluded from the instrumentation. Also agent does not include grpc libraries, you can see the list of libraries that are included in the agent in https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/licenses/licenses.md

**Reply by huange7** [reply]
I'm sorry I don't know enough about the agent. 
For example, in the UpstreamGrpcExporter class, the agent uses grpc for upstreaming. The related dependency is also referenced in build.gradle.kts, won't this cause the agent to introduce this dependency?

**Reply by laurit** [reply]
Agent used to include the grpc libraries but that was changed ~1.5 years ago. I guess you are looking at an old version?

**Reply by huange7** [reply]
Sorry, forgot to mention the agent version. I'm looking at the 1.16 version of the code. But here's how previous agent solved the problem?

**Reply by laurit** [reply]
> Most of the agent code is loaded into separate class loader that is excluded from the instrumentation.

**Reply by huange7** [reply]
Here I don't really understand, if they are loaded by different class loaders, won't they be transformed?

**Reply by laurit** [reply]
https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/e4445ab507d8c0c76289df7ab7c32acb9d627547/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/ignore/GlobalIgnoredTypesConfigurer.java#L127

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/e4445ab507d8c0c76289df7ab7c32acb9d627547/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/ignore/IgnoredClassLoadersMatcher.java#L38-L43

**Reply by huange7** [reply]
Oh! Thanks for your guidance.
